ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthConfig.java

```diff
@@ -0,0 +1,119 @@
+import org.hyperledger.besu.ethereum.mainnet.requests.RequestProcessorCoordinator;
+import org.hyperledger.besu.evm.precompile.PrecompileContractRegistry;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Supplier;
+import java.util.zip.CRC32;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Suppliers;
+
+public class EthConfig implements JsonRpcMethod {
+
+  private static final Supplier<ObjectMapper> mapperSupplier = Suppliers.memoize(ObjectMapper::new);
+
+  private final BlockchainQueries blockchain;
+  private final ProtocolSchedule protocolSchedule;
+
+  public EthConfig(final BlockchainQueries blockchain, final ProtocolSchedule protocolSchedule) {
+    this.blockchain = blockchain;
+    this.protocolSchedule = protocolSchedule;
+  }
+
+  @Override
+  public String getName() {
+    return RpcMethod.ETH_CONFIG.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    var header = blockchain.getBlockchain().getChainHeadHeader();
+    var current = protocolSchedule.getForNextBlockHeader(header, System.currentTimeMillis() / 1000);
+    var next = protocolSchedule.getNextProtocolSpec(current);
+
+    var currentConfig = generateConfig(current);
+    var nextConfig = next.map(this::generateConfig).orElse(null);
+
+    ObjectNode result = mapperSupplier.get().createObjectNode();
+    result.put("current", currentConfig);
+    result.put("currentHash", configHash(currentConfig));
+    result.put("next", nextConfig);
+    result.put("nextHash", configHash(nextConfig));
+
+    return new JsonRpcSuccessResponse(requestContext.getRequest().getId(), result);
+  }
+
+  JsonNode generateConfig(final ProtocolSpec spec) {
+    var result = mapperSupplier.get().createObjectNode();
+    var forkId = protocolSchedule.hardforkFor(x -> x.spec() == spec).orElseThrow();
+
+    result.put("activation", forkId.milestone());
+
+    var blobs = result.putObject("blobs");
+    blobs.put(
+        "baseFeeUpdateFraction", spec.getFeeMarket().getBaseFeeUpdateFraction().longValueExact());
+    blobs.put("max", spec.getGasLimitCalculator().currentBlobGasLimit() / (128 * 1024));
+    blobs.put("target", spec.getGasCalculator().getBlobTarget());
+
+    result.put(
+        "chainId", protocolSchedule.getChainId().map(c -> "0x" + c.toString(16)).orElse(null));
+
+    var contracts =
+        new TreeMap<>(
+            spec.getRequestProcessorCoordinator()
+                .map(RequestProcessorCoordinator::getContractConfigs)
+                .orElse(Map.of()));
+    spec.getBlockHashProcessor()
+        .getHistoryContract()
+        .ifPresent(a -> contracts.put("HISTORY", a.toHexString()));
+    if (!contracts.isEmpty()) {
+      var jsonContracts = result.putObject("contracts");
```

Comment:
```suggestion
      var jsonContracts = result.putObject("systemContracts");
```
to distinguish from other pre-deployed contracts that might be in the genesis